packages/engine-core/src/binary/BinaryEngine.ts

```diff
@@ -1110,66 +1130,70 @@ You very likely have the wrong "binaryTarget" defined in the schema.prisma file.
     })
   }
 
-  private handleRequestError = async (error: Error & { code?: string }, graceful = false) => {
+  private handleRequestError = async (
+    error: Error & { code?: string },
+  ): Promise<{ error: Error & { code?: string }; shouldRetry: boolean }> => {
     debug({ error })
+
     // if we are starting, wait for it before we handle any error
     if (this.startPromise) {
       await this.startPromise
     }
 
+    // matching on all relevant error codes from
+    // https://github.com/nodejs/undici/blob/2.x/lib/core/errors.js
+    const isNetworkError = [
+      'ECONNRESET',
+      'ECONNREFUSED',
+      'UND_ERR_CLOSED',
+      'UND_ERR_SOCKET',
+      'UND_ERR_DESTROYED',
+      'UND_ERR_ABORTED',
+    ].includes(error.code as string)
+
     if (error instanceof PrismaClientKnownRequestError) {
-      throw error
+      return { error, shouldRetry: false }
     }
 
-    this.throwAsyncErrorIfExists()
-
-    // A currentRequestPromise is only being canceled by the sendPanic function
-    if (this.currentRequestPromise?.isCanceled) {
+    try {
       this.throwAsyncErrorIfExists()
-    } else if (
-      // matching on all relevant error codes from
-      // https://github.com/nodejs/undici/blob/2.x/lib/core/errors.js
-      error.code === 'ECONNRESET' ||
-      error.code === 'ECONNREFUSED' ||
-      error.code === 'UND_ERR_CLOSED' ||
-      error.code === 'UND_ERR_SOCKET' ||
-      error.code === 'UND_ERR_DESTROYED' ||
-      error.code === 'UND_ERR_ABORTED' ||
-      error.message.toLowerCase().includes('client is destroyed') ||
-      error.message.toLowerCase().includes('other side closed') ||
-      error.message.toLowerCase().includes('the client is closed')
```

Comment:
I think these three (now removed) also refer to some kind of network errors, but yeah, might be safer not to match them to avoid false positives.

Reply:
We now check only the codes, this felt like a double-check for no reason

https://github.com/nodejs/undici/blob/2.x/lib/core/errors.js